### PR TITLE
[TT-11470] Add human identifiable information in NodeData

### PR DESCRIPTION
**Pull request by padiazg**
## Description

Provide IP, PID and Hostname as identifiers for a node when querying MDCB in the `dataplanes` endpoint. This should be tested with the MDCB version in [this pr](https://github.com/TykTechnologies/tyk-sink/pull/525) 

## Related Issue

https://tyktech.atlassian.net/browse/TT-11470

## How This Has Been Tested

- Run Tyk in MDCB Env
- Enable secure endpoints in MDCB
- Consume the `/dataplanes` endpoint, and now you get host details per node

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Refactoring or add test (improvements in base code or adds test coverage to functionality)

## Checklist

- [ ] I ensured that the documentation is up to date
- [ ] I explained why this PR updates go.mod in detail with reasoning why it's required
- [ ] I would like a code coverage CI quality gate exception and have explained why

___

## **Type**
enhancement, tests

___

## **Description**
- Implemented a buffered logger (`BufferedLogger` and `BufferingFormatter`) to facilitate better logging, especially for testing scenarios.
- Enhanced the `Gateway` struct to include node IP address handling, which fetches and stores the IP if not provided.
- Added comprehensive tests for new features including the retrieval and logging of node IP addresses.
- Introduced a utility function `getIpAddress` to fetch the first non-loopback IPv4 address, enhancing node identification.

___

## **Changes walkthrough**
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement
</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>buffered-logger.go</strong><dd><code>Implement Buffered Logger for Enhanced Testing</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

gateway/buffered-logger.go
<li>Added a new <code>BufferedLogger</code> and <code>BufferingFormatter</code> to handle logging in <br>a buffered manner, primarily for use in tests.<br> <li> <code>BufferedLogger</code> includes methods to retrieve logs of a specific level.<br>

</details>
    

  </td>
  <td><a href="https://github.com/TykTechnologies/tyk/pull/6229/files#diff-6f5a2cb4531e50ec9d0929cb7c1e5437c72dc829f5a1fe722048b0b117f391a2">+72/-0</a>&nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>server.go</strong><dd><code>Enhance Gateway with Node IP Address Handling</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

gateway/server.go
<li>Added <code>Address</code> field to <code>hostDetails</code> struct to store node IP address.<br> <li> Enhanced <code>getHostDetails</code> to fetch and store the node's IP address if <br>not already provided.<br>

</details>
    

  </td>
  <td><a href="https://github.com/TykTechnologies/tyk/pull/6229/files#diff-4652d1bf175a0be8f5e61ef7177c9666f23e077d8626b73ac9d13358fa8b525b">+9/-1</a>&nbsp; &nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>util.go</strong><dd><code>Add Utility Function to Fetch Non-Loopback IPv4 Address</code>&nbsp; &nbsp; </dd></summary>
<hr>

gateway/util.go
<li>Added <code>getIpAddress</code> function to retrieve the first non-loopback IPv4 <br>address.<br>

</details>
    

  </td>
  <td><a href="https://github.com/TykTechnologies/tyk/pull/6229/files#diff-1aa619f406837fb44ac6fba2c8922795eba0285dc4c8d2b0c15755b60b9ee1a5">+25/-0</a>&nbsp; &nbsp; </td>
</tr>                    
</table></td></tr><tr><td><strong>Tests
</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>server_test.go</strong><dd><code>Add Tests for Gateway Host Details and IP Address Retrieval</code></dd></summary>
<hr>

gateway/server_test.go
<li>Added tests for <code>getHostDetails</code> to verify correct logging and IP <br>address retrieval.<br> <li> Utilized <code>BufferedLogger</code> in tests to capture log outputs.<br>

</details>
    

  </td>
  <td><a href="https://github.com/TykTechnologies/tyk/pull/6229/files#diff-d9f006370c9748c09affd99d0a4edeb8f3419057703a67fd70838a764a485696">+109/-2</a>&nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>util_test.go</strong><dd><code>Implement Tests for IP Address Retrieval Utility Function</code></dd></summary>
<hr>

gateway/util_test.go
<li>Added tests for <code>getIpAddress</code> to ensure it correctly identifies and <br>returns non-loopback IPv4 addresses.<br>

</details>
    

  </td>
  <td><a href="https://github.com/TykTechnologies/tyk/pull/6229/files#diff-3ca88235eab68f9eb691c139bb6e45f10a4038d4c02cf435f66ef7394734f925">+82/-1</a>&nbsp; &nbsp; </td>
</tr>                    
</table></td></tr></tr></tbody></table>

___

> ✨ **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

### CODE REVIEW

File: gateway/buffered-logger.go [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+import (
+	"sync"
+	"time"
+
+	"github.com/sirupsen/logrus"
+)
+
+// implements a buffered logger mainly to be used in tests
+
+type BufferedLog struct {
```
Comment (titpetric) [inline gateway/buffered-logger.go:12]: Please move to `internal/log` when it's ready :)

File: gateway/util.go [hunk 2]
```diff
@@ -216,3 +216,28 @@ func containsEscapedChars(str string) bool {
...
 }
+
+// getIpAddress returns the first non-loopback IPv4 address found. Returns error if it fails
+// to get the list of addresses. Returns empty if there's no valid IP addresses.
+// netInterfaceAddrs is used to allow mocking in the tests
+var (
+	netInterfaceAddrs = net.InterfaceAddrs
+)
+
+func getIpAddress() (string, error) {
```
Comment (titpetric) [inline gateway/util.go:227]: This also needs a move, and possibly ipv6 support

@jeffy-mathew internal/netutil, wdyt?

**Comment by titpetric** [comment]
Hey @padiazg, nice to see a contribution from your side. Could you please consider moving any extractable parts into subpackages of /internal? The PR:

- extends internal apis
- adds a buffered logger
- modifies a gateway function that could be extracted

if there's no better location, create an `internal/service` for the new symbols. Ideally all gateway contributions should work to not increase the size of the gateway package, separating the concerns and ensuring low coupling in a tight scope.

File: gateway/util.go (hunk 2 above)
Comment (titpetric) [reply gateway/util.go:227]: Also, hosts may have multiple IP addresses, needs to be a form of []string as it's usually not clear which one is a "primary".

Comment (padiazg) [reply gateway/util.go:227]: Yes, but for the need of the ticket we only need one valid IP, we ignore loopback ones.

Comment (padiazg) [reply gateway/util.go:227]: > This also needs a move
indeed, I was wondering where it would fit

File: gateway/buffered-logger.go (hunk 1 above)
Comment (padiazg) [reply gateway/buffered-logger.go]: done

**Comment by padiazg** [comment]
> * adds a buffered logger

this is meant to be used only to catch the logs during tests, not to be used in any regular flow

File: internal/netutil/ip-address_test.go [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+		},
+	}
+	for _, tt := range tests {
+		tt := tt
+		t.Run(tt.name, func(t *testing.T) {
+			// use mocked net.InterfaceAddrs
+			netInterfaceAddrs = tt.netInterfaceAddrs
+
+			got, err := GetIpAddress()
+			if (err != nil) != tt.wantErr {
```
Comment (titpetric) [inline internal/netutil/ip-address_test.go:78]: Use testify/assert.

File: internal/netutil/ip-address.go [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+import "net"
+
+// getIpAddress returns the first non-loopback IPv4 address found. Returns error if it fails
+// to get the list of addresses. Returns empty if there's no valid IP addresses.
+// netInterfaceAddrs is used to allow mocking in the tests
+var (
+	netInterfaceAddrs = net.InterfaceAddrs
+)
+
+func GetIpAddress() ([]string, error) {
```
Comment (titpetric) [inline internal/netutil/ip-address.go:12]: Fix godoc :D ...

File: internal/log/buffered-logger.go [hunk 5]
```diff
@@ -0,0 +1,77 @@
+package gateway
```
Comment (titpetric) [inline internal/log/buffered-logger.go:1]: Use `_` in go filenames, following existing go conventions. Fix all files.

File: internal/log/buffered-logger.go [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+import (
+	"sync"
+	"time"
+
+	"github.com/sirupsen/logrus"
+)
+
+// implements a buffered logger mainly to be used in tests
+
+type BufferedLog struct {
```
Comment (titpetric) [inline internal/log/buffered-logger.go:12]: Cover with godoc (golangci-lint run ./... would report these)

File: gateway/server_test.go [hunk 7]
```diff
@@ -216,3 +220,144 @@ func TestGateway_SyncResourcesWithReload(t *testing.T) {
...
 }
+
+func TestGateway_getHostDetails(t *testing.T) {
+	type checkFn func(*testing.T, *il.BufferedLogger, *Gateway)
+
+	var (
+		orig_readPIDFromFile = readPIDFromFile
+		orig_mainLog         = mainLog
+		orig_getIpAddress    = netutil.GetIpAddress
+		bl                   = il.NewBufferingLogger()
```
Comment (titpetric) [inline gateway/server_test.go:231]: Should the buffering logger just be a test utility? In that case I misled you, and the log utilities should live in `/test/log.go`. I apologize for the oversight.

File: internal/netutil/ip_address_test.go [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+
+import (
+	"fmt"
+	"net"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func Test_getIpAddress(t *testing.T) {
```
Comment (titpetric) [inline internal/netutil/ip_address_test.go:11]: Naming, Test_GetIpAddress (match symbol case sensitive)

File: internal/models/host_details.go [hunk 9]
```diff
@@ -0,0 +1,9 @@
+package models
+
+// HostDetails contains information about a host machine,
+// including its hostname, process ID (PID), and IP address.
+type HostDetails struct {
```
[the remark this answers is not in the document]
Comment (sredxny) [reply internal/models/host_details.go:5]: issue is that `HostDetails` is being imported by `apidef/rpc.go` and `apidef` is being imported by the gateway pkg, which generated a cyclic import. The other way is to move `HostDetails` to the `apidef` pkg but it doesn't looks right as they are totally different things

File: gateway/server.go [hunk 10]
```diff
@@ -1540,6 +1540,17 @@ func (gw *Gateway) getHostDetails(file string) {
 	if gw.hostDetails.Hostname, err = os.Hostname(); err != nil {
 		mainLog.Error("Failed to get hostname: ", err)
 	}
+
+	gw.hostDetails.Address = gw.GetConfig().ListenAddress
+	if gw.hostDetails.Address == "" {
```
[the remark this answers is not in the document]
Comment (padiazg) [reply gateway/server.go:1545]: probably, but that was on the original code: https://github.com/TykTechnologies/tyk/blob/master/gateway/server.go#L1923
I just changed `gw.GetConfig().ListenAddress` for `gw.hostDetails.Address` not to call the function twice.

File: gateway/server_test.go [hunk 11]
```diff
@@ -216,3 +221,161 @@ func TestGateway_SyncResourcesWithReload(t *testing.T) {
...
+	netutilGetIpAddress func() ([]string, error)
+	checks              []gatewayGetHostDetailsTestCheckFn
+	// config              config.Config
+} {
+	var (
+		check = func(fns ...gatewayGetHostDetailsTestCheckFn) []gatewayGetHostDetailsTestCheckFn { return fns }
+		// matches ipv6 and ipv4
+		// https://go.dev/play/p/cbOQUiqNqyU
+		// https://regex101.com/r/lIWfaA/1
+		ipAddrPattern = `^((([a-f0-9]{1,4}):){7}([a-f0-9]{1,4})|(([a-f0-9]{1,4})(:([a-f0-9]{1,4})){0,6})?::(([a-f0-9]{1,4})(:([a-f0-9]{1,4})){0,6})?|((([a-f0-9]{1,4}):){5}[a-f0-9]{1,4}|(([a-f0-9]{1,4}):){0,5}:([a-f0-9]{1,4}))|(([a-f0-9]{1,4}:){0,6}[a-f0-9]{1,4}|(([a-f0-9]{1,4}:){0,6}:([a-f0-9]{1,4})){0,1})|([a-f0-9]{1,4}:){0,7}:|([a-f0-9]{1,4}:){0,6}[a-f0-9]{1,4}|(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?))$`
```
[the remark this answers is not in the document]
Comment (padiazg) [reply gateway/server_test.go:273]: what would be an alternative to validate if the address is either a valid IPv4 or IPv6 address?

[the remark this answers is not in the document]
Comment (padiazg) [reply gateway/server_test.go:273]: what would be an alternative to validate if the address is either a valid IPv4 or IPv6 address?

File: gateway/server_test.go [hunk 12]
```diff
@@ -216,3 +221,161 @@ func TestGateway_SyncResourcesWithReload(t *testing.T) {
...
 
 }
+
+type gatewayGetHostDetailsTestCheckFn func(*testing.T, *test.BufferedLogger, *Gateway)
+
+func gatewayGetHostDetailsTestHasErr(wantErr bool, errorText string) gatewayGetHostDetailsTestCheckFn {
+	return func(t *testing.T, bl *test.BufferedLogger, _ *Gateway) {
+		t.Helper()
+		logs := bl.GetLogs(logrus.ErrorLevel)
+		if !wantErr && assert.Empty(t, logs) {
```
[the remark this answers is not in the document]
Comment (padiazg) [reply gateway/server_test.go]: changed

File: internal/netutil/ip_address_test.go (hunk 8 above)
Comment (padiazg) [reply internal/netutil/ip_address_test.go]: fixed

**Review by titpetric — APPROVED** [review]
Mostly reviewed for the code structure conventions, models/ needs a rename, everything else LGTM

**Review by sredxny — APPROVED** [review]
looks good! Only address the comment above from Tit

**Pull request merged**